### DISCUSSION TITLE
Xdrip+ on Android 4.4.2

### DISCUSSION THREAD

**Discussion by renegadeandy** [NightscoutFoundation/xDrip#3700] [Q&A] [answered]
Hi, It looks like XDrip+ supports android 4.4.2 

I have an old tablet I want to use as a follower for XDrip+, so I downloaded the apk, and when installing it says "Error parsing package". I Validated that an old version of XDrip+ from 2016 will install successfully.

Given the latest changes to followers, would an old old version of XDrip+ still work as a follower on this device, or do I really need the very latest.

If I need the very latest, can somebody help me figure out how to get this installed given it should be supported?

**Comment by renegadeandy** [comment]
Doing an ADB install from the laptop shows:

Performing Push Install
xDrip-plus-20240917-dfcbe80.apk: 1 file pushed, 0 skipped. 6.0 MB/s (15996657 bytes in 2.552s)
	pkg: /data/local/tmp/xDrip-plus-20240917-dfcbe80.apk
Failure [INSTALL_FAILED_OLDER_SDK]

**Comment by renegadeandy** [comment]
It looks to me as if the app requires android SDK 24 -- which is Android 7?

So is this page wrong : https://xdrip.readthedocs.io/en/latest/install/prerequisites/ ?

Do you think I could compile it for android sdk : 19 - which would work on android sdk 4.4?

**Comment by Navid200** [comment]
Android 6 was retired on May 29, 2024:  https://navid200.github.io/xDrip/docs/ReleaseNotes.html

**Comment by renegadeandy** [comment]
Hi @Navid200. Okay - so the link in the docs should be updated too.

Do you think it would be doable to take the pre may build, incorporate the fix for Follower, and make a build 4.4 compatible. I think it would work, and would be somewhat simple to achieve. Thoughts?

**Reply by Navid200** [reply]
I don't see why not if you mean for your private version.  You can do whatever you want.

**Reply by Navid200** [reply]
But, that would make it compatible with Android 6, not 5 or 4.

**Reply by Navid200** [reply]
Android 5 was retired on June 15, 2023.
Android 4 was retired on October 25, 2022.

**Comment by renegadeandy** [comment]
Hi @Navid200 I am having trouble tracking down what code changes were tied to the follower fix on the recent update, (so I can back port them to my older base xDrip+).

Can you please link me to the commit where the fix was included for Sync Follower?

**Reply by Navid200** [reply]
The first change was made on July 23 that changed the behavior of xDrip Sync to address the recent changes made by Google.
Then, there were three releases made on September 15 that made more changes along those lines.

But, we have moved to AndroidX at the end 2023.  I am not sure if you can just take a few commits and apply them to a release from 2022.

Are you doing all of this so that you can use Android 4?  Isn't Android 4 really old and outdated?

**Reply by renegadeandy** [reply]
Yeah it is old. And yeah I'm doing this to support Android 4.4.2 for 2 reasons:

1/ That's the device And OS I have and seemingly unable to upgrade it as this is my only spare device and I would like to utilise  it instead of wasting it.

2/ The xdrip docs stated 4.4 was supported and so now I have a kind of challenge to make my intention work.

**Reply by Navid200** [reply]
I don't think there is any claim that xDrip is compatible with Android 4 anywhere in this repository. 

Anyone can put together a guide and call it xDrip documentation.  There is no way we can monitor/correct/update all of those.
You can still find guides online that tell you to enable engineering mode if you want to hard reset a Dexcom transmitter.  That capability was removed from engineering mode in 2018.  There is so much incorrect info about xDrip all over the place.
  
If there is a doc in this repository that is invalid, I would appreciate a pointer so that we can address it.

Thanks and hope it works out

**Comment by renegadeandy** [comment]
@Navid200 : I posted this 2 days ago:

https://github.com/NightscoutFoundation/xDrip/discussions/3700#discussioncomment-10853959

**Reply by Navid200** [reply]
And I saw it.  It is a link to a guide someone put together.  
It is not a link to anything in this repository.

**Reply by renegadeandy** [reply]
Well I suppose we can all ignore it then and let others fall into the same traps as me. Good.

I suppose another question to ask is, will the fixes to follower be compatible with 4.4.2. Because it is so hard to tie down the exact changes for this feature, answering this question is really difficult.

**Reply by Navid200** [reply]
I cannot answer that question.

Our focus now is Android 7-15.
When Android 15 beta came out, we had no choice but to retire Android 6.  When something like that happens, we have no choice.  
I don't want to force people to buy new phones.  I hate doing that.  But, unless we release our own operating system, we are limited by what Google imposes on us.

**Comment by renegadeandy** [comment, marked as answer]
Well - for everyone that is interested I actually got it to work! Steps were:

Enable multidex in gradle build files manually (as 4.4.2 it wasn't standard)
Change target and minimum sdk to level 19
Keep build level at 34

I hit 1000 snags along the way but I ended up with xDrip+ on my old 4.4.2 tablet, AND what's even better is the xDrip+ Companion is syncing up perfectly! It works a charm!

(When opening the app is sometimes crashes which I think is to do with memory, but a couple of open attempts and it opens fine, at which point it is very stable!)
